### PR TITLE
feat: per-user per-model chat compaction threshold overrides

### PR DESCRIPTION
**Pull request by ibetitsmike**
## What

Adds per-user per-model auto-compaction threshold overrides. Users can now customize the percentage of context window usage that triggers chat compaction, independently for each enabled model.

## Why

The compaction threshold was previously only configurable at the deployment level (`chat_model_configs.compression_threshold`). Different users have different preferences — some want aggressive compaction to keep costs low, others prefer higher thresholds to retain more context. This gives users control without requiring admin intervention.

## Architecture

**Storage:** Reuses the existing `user_configs` table (no migration needed). Overrides are stored as key/value pairs with keys shaped `chat_compaction_threshold:<modelConfigID>` and integer percent values.

**API:** Three new experimental endpoints under `/api/experimental/chats/config/`:
- `GET /user-compaction-thresholds` — list all overrides for the current user
- `PUT /user-compaction-thresholds/{modelConfig}` — upsert an override (validates model exists and is enabled, validates 0–100 range)
- `DELETE /user-compaction-thresholds/{modelConfig}` — clear an override (idempotent)

**Runtime resolution:** In `coderd/chatd/chatd.go`, a new `resolveUserCompactionThreshold()` helper runs at the start of each chat turn (inside `runChat()`), after the model config is resolved but before `CompactionOptions` is built. If a valid override exists, it replaces `modelConfig.CompressionThreshold`. The threshold source (`user_override` vs `model_default`) is logged with each compaction event.

**Precedence:** `effectiveThreshold = userOverride ?? modelConfig.CompressionThreshold`

**UI:** New "Context Compaction" subsection in the Agents → Settings → Behavior tab, placed after Personal Instructions. Shows one row per enabled model with the system default, a number input for the override, and Save/Reset controls.

## Testing

- 9 API subtests covering CRUD, validation (boundary values 0/100, out-of-range rejection), upsert behavior, idempotent delete, user isolation, and non-existent model config
- 4 dbauthz tests (16 scenarios) verifying `ActionReadPersonal` / `ActionUpdatePersonal` on all query methods
- 4 Storybook stories with play functions (Default, WithOverrides, Loading, Error)

<details>
<summary>Implementation plan</summary>

### Phase 1 — Tests
- Backend API tests in `coderd/chats_test.go` (9 subtests)
- Database auth wrapper tests in `coderd/database/dbauthz/dbauthz_test.go` (4 methods)
- Frontend stories in `UserCompactionThresholdSettings.stories.tsx` (4 stories)

### Phase 2 — Backend preference surface
- 4 SQL queries in `coderd/database/queries/users.sql` (list, get, upsert, delete)
- `make gen` to propagate into generated artifacts
- Auth/metrics wrappers in dbauthz and dbmetrics
- SDK types and client methods in `codersdk/chats.go`
- HTTP handlers and routes in `coderd/chats.go` and `coderd/coderd.go`
- Key prefix constant shared between handlers and runtime

### Phase 3 — Runtime override
- `resolveUserCompactionThreshold()` helper in `coderd/chatd/chatd.go`
- Override injection in `runChat()` before building `CompactionOptions`
- `threshold_source` field added to compaction log

### Phase 4 — Settings UI
- API client methods and React Query hooks in `site/src/api/`
- `UserCompactionThresholdSettings` component extracted from `SettingsPageContent`
- Per-model mutation tracking (only the active row disables during save)
- 100% warning, "System default" label, helpful empty state copy

### Phase 5 — Refactor and review fixes
- Consolidated key prefix constant in `codersdk`
- Explicit PUT range validation (not just struct tags)
- GET handler gracefully skips malformed rows instead of 500
- Boundary value, upsert, and non-existent model config tests
- UX improvements: per-model mutation state, aria-live on errors

</details>

### CODE REVIEW

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

File: site/src/pages/AgentsPage/UserCompactionThresholdSettings.tsx [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+														onError: (error) => {
+															setRowErrors((currentErrors) => ({
+																...currentErrors,
+																[modelConfig.id]: getErrorMessage(
+																	error,
+																	"Failed to reset compaction threshold.",
+																),
+															}));
+														},
+													});
```
Comment (DanielleMaywood) [inline site/src/pages/AgentsPage/UserCompactionThresholdSettings.tsx]: You can define these `onSuccess` and `onError` handlers at the actual `useMutation`. It'll make the markup nicer to read.

**Review by DanielleMaywood — COMMENTED** [review]
Unsure on the best practices so converse with your agent, but it _feels_ like I'm expecting to see a `<form>` somewhere but I'm not.

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

File: coderd/database/queries/users.sql [hunk 2]
```diff
@@ -193,6 +193,25 @@ WHERE user_configs.user_id = @user_id
 	AND user_configs.key = 'chat_custom_prompt'
 RETURNING *;
 
+-- name: ListUserChatCompactionThresholds :many
+SELECT user_id, key, value FROM user_configs
+WHERE user_id = @user_id AND key LIKE 'chat_compaction_threshold:%'
```
Comment (johnstcn) [inline coderd/database/queries/users.sql]: Should include the unit for clarity e.g. `chat_compaction_threshold_pct`

File: coderd/chats.go [hunk 3]
```diff
@@ -2816,6 +2831,170 @@ func (api *API) putUserChatCustomPrompt(rw http.ResponseWriter, r *http.Request)
...
+			continue
+		}
+		if thresholdPercent < int64(minChatContextCompressionThreshold) ||
+			thresholdPercent > int64(maxChatContextCompressionThreshold) {
+			api.Logger.Warn(ctx, "skipping out-of-range user chat compaction threshold",
+				slog.F("key", row.Key),
+				slog.F("value", row.Value),
+			)
+			continue
+		}
```
Comment (johnstcn) [inline coderd/chats.go:2880]: suggestion: clamp to [min,max] threshold

File: coderd/chats.go [hunk 4]
```diff
@@ -2816,6 +2831,170 @@ func (api *API) putUserChatCustomPrompt(rw http.ResponseWriter, r *http.Request)
...
+		if err != nil {
+			api.Logger.Warn(ctx, "skipping malformed user chat compaction threshold value",
+				slog.F("key", row.Key),
+				slog.F("value", row.Value),
+				slog.Error(err),
+			)
+			continue
+		}
+		if thresholdPercent < int64(minChatContextCompressionThreshold) ||
+			thresholdPercent > int64(maxChatContextCompressionThreshold) {
```
Comment (johnstcn) [inline coderd/chats.go:2874]: Less than or less than/equal to? Are you allowed set min/max threshold?

File: coderd/exp_chats.go [hunk 5]
```diff
@@ -2542,6 +2542,21 @@ func normalizeChatCompressionThreshold(
 	return threshold, nil
 }
 
+func compactionThresholdKey(modelConfigID uuid.UUID) string {
+	return codersdk.ChatCompactionThresholdKeyPrefix + modelConfigID.String()
+}
+
+func parseCompactionThresholdKey(key string) (uuid.UUID, error) {
```
Comment (johnstcn) [inline coderd/exp_chats.go:2545]: Wouldn't it also make sense here to parse and return the value as an integer?

File: coderd/chats_test.go [hunk 6]
```diff
@@ -4926,6 +4926,183 @@ func TestChatWorkspaceTTL(t *testing.T) {
...
 }
 
+func TestUserChatCompactionThresholds(t *testing.T) {
+	t.Parallel()
+
+	t.Run("EmptyByDefault", func(t *testing.T) {
+		t.Parallel()
+
+		ctx := testutil.Context(t, testutil.WaitLong)
+		client := newChatClient(t)
```
Comment (johnstcn) [inline coderd/chats_test.go]: just create one coderdtest instance and then create / read / update / delete in serial - saves on database instance.

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

File: coderd/x/chatd/chatd.go [hunk 7]
```diff
@@ -3715,6 +3723,26 @@ func (p *Server) resolveInstructions(
...
+	if err != nil {
+		// sql.ErrNoRows is the normal "not set" case.
+		return 0, false
+	}
+	val, err := strconv.ParseInt(raw, 10, 32)
+	if err != nil || val < 0 || val > 100 {
+		return 0, false
+	}
+	return int32(val), true
+}
```
Comment (johnstcn) [inline coderd/x/chatd/chatd.go:3755]: suggestion, non-blocking: It feels like it would be easy for a future editor to forget to apply the user's compaction override somewhere else. Wouldn't it be more ergonomic here to return the compaction threshold for the model if there is no user-level override? Say, in `p.resolveConfig`?

**Review by johnstcn — APPROVED** [review]
Backend LGTM 👍

**Comment by ibetitsmike** [comment]
@codex review

File: site/src/pages/AgentsPage/AgentDetail.tsx [hunk 8]
```diff
@@ -495,10 +497,17 @@ const AgentDetail: FC = () => {
...
+		if (!chatLastModelConfigID) return undefined;
+		const userOverride = userThresholdsQuery.data?.thresholds.find(
+			(t) => t.model_config_id === chatLastModelConfigID,
+		);
+		if (userOverride) {
+			return userOverride.threshold_percent;
+		}
+		return modelConfigs.find((c) => c.id === chatLastModelConfigID)
+			?.compression_threshold;
+	})();
```
Comment (DanielleMaywood) [inline site/src/pages/AgentsPage/AgentDetail.tsx]: Maybe worth moving this logic into a utility function, what do you think?

**Review by DanielleMaywood — APPROVED** [review]

**Comment by ibetitsmike** [comment]
@codex review

File: coderd/x/chatd/chatd.go [hunk 9]
```diff
@@ -3150,8 +3151,14 @@ func (p *Server) runChat(
 	// "Summarizing..." tool call with the "Summarized" tool
 	// result.
 	compactionToolCallID := "chat_summarized_" + uuid.NewString()
+	effectiveThreshold := modelConfig.CompressionThreshold
+	thresholdSource := "model_default"
+	if override, ok := p.resolveUserCompactionThreshold(ctx, chat.OwnerID, modelConfig.ID); ok {
+		effectiveThreshold = override
```
Comment (mafredri) [inline coderd/x/chatd/chatd.go:3157]: 🤖 Agent on behalf of mafredri

**P1** The feature's core behavioral contract is untested. *(Bisky P1, Meruem P2)*

> `resolveUserCompactionThreshold` is called in `runChat` to override `modelConfig.CompressionThreshold` with a user-set value. This is the only place where user compaction thresholds actually affect chat behavior. Grepping `coderd/` test files for `resolveUserCompactionThreshold` or `resolveCompactionThreshold` returns zero hits. [...] If someone deleted the six lines in `runChat` that call `resolveUserCompactionThreshold` and wire `effectiveThreshold` into `CompactionOptions`, every test would still pass. That's the definition of a test that tests the plumbing but not the feature.

The CRUD tests prove the API endpoints work, but nothing verifies that a user setting threshold 50 on a model with default 80 actually results in `CompactionOptions.ThresholdPercent == 50` during chat execution. A test that sets an override, runs a chat turn, and asserts the effective threshold would close this gap. This is the one finding I'd want addressed before merge.

File: coderd/database/dbauthz/dbauthz_test.go [hunk 10]
```diff
@@ -2278,6 +2278,35 @@ func (s *MethodTestSuite) TestUser() {
 		dbm.EXPECT().UpdateUserChatCustomPrompt(gomock.Any(), arg).Return(uc, nil).AnyTimes()
 		check.Args(arg).Asserts(u, policy.ActionUpdatePersonal).Returns(uc)
 	}))
+	s.Run("ListUserChatCompactionThresholds", s.Mocked(func(dbm *dbmock.MockStore, faker *gofakeit.Faker, check *expects) {
+		u := testutil.Fake(s.T(), faker, database.User{})
+		uc := database.UserConfig{UserID: u.ID, Key: "chat_compaction_threshold:00000000-0000-0000-0000-000000000001", Value: "75"}
```
Comment (mafredri) [inline coderd/database/dbauthz/dbauthz_test.go]: 🤖 Agent on behalf of mafredri

**P2** All four dbauthz test cases use the wrong key prefix. *(Kurapika P1, Mafu-san P1, Hisoka P2, Mafuuu P2)*

> The tests hardcode `"chat_compaction_threshold:00000000-..."` (no `_pct` suffix). The production constant is `ChatCompactionThresholdKeyPrefix = "chat_compaction_threshold_pct:"`. The SQL LIKE pattern also uses `chat\_compaction\_threshold\_pct:`. The mock-based tests only verify auth policy assertions, so the wrong prefix doesn't cause failures today. But these tests provide false confidence: they'd still pass if the production prefix drifted, and anyone reading the tests gets a wrong mental model of the key format.

Fix: use `codersdk.ChatCompactionThresholdKeyPrefix + "00000000-0000-0000-0000-000000000001"` to keep the test keys in sync with production. This affects all four new test cases (`ListUserChatCompactionThresholds`, `GetUserChatCompactionThreshold`, `UpdateUserChatCompactionThreshold`, `DeleteUserChatCompactionThreshold`).

File: coderd/chatd/chatd.go [hunk 11]
```diff
@@ -3715,6 +3723,26 @@ func (p *Server) resolveInstructions(
...
+// compaction threshold override. Returns the override value and
+// true if one exists and is valid, or 0 and false otherwise.
+func (p *Server) resolveUserCompactionThreshold(ctx context.Context, userID uuid.UUID, modelConfigID uuid.UUID) (int32, bool) {
+	key := codersdk.ChatCompactionThresholdKeyPrefix + modelConfigID.String()
+	raw, err := p.db.GetUserChatCompactionThreshold(ctx, database.GetUserChatCompactionThresholdParams{
+		UserID: userID,
+		Key:    key,
+	})
+	if err != nil {
+		// sql.ErrNoRows is the normal "not set" case.
```
Comment (mafredri) [inline coderd/chatd/chatd.go]: 🤖 Agent on behalf of mafredri

**P2** `resolveUserCompactionThreshold` silently swallows all DB errors, not just `sql.ErrNoRows`. *(Knuckle P2, Meruem P2, Luffy P2)*

> The function catches `err != nil` from `GetUserChatCompactionThreshold` and returns `0, false` for every error type. A real DB failure (connection timeout, pool exhaustion) silently falls back to the model default. For a user who set threshold to 100% (disable compaction), a transient DB error would re-enable compaction without any log line.

The comment acknowledges `ErrNoRows` is the normal case, but the code doesn't distinguish it from operational failures. The existing `resolveUserPrompt` has the same pattern, so this is consistent — but the fix is a one-liner:

```go
if err != nil {
    if !errors.Is(err, sql.ErrNoRows) {
        p.logger.Warn(ctx, "failed to resolve user compaction threshold",
            slog.F("user_id", userID),
            slog.F("model_config_id", modelConfigID),
            slog.Error(err),
        )
    }
    return 0, false
}
```

This preserves the graceful fallback while making DB health issues visible in logs.

File: AGENTS.md [hunk 12]
```diff
@@ -219,8 +219,11 @@ seems like it should use `time.Sleep`, read through https://github.com/coder/qua
 - Commit format: `type(scope): message`
 - PR titles follow the same `type(scope): message` format.
 - When you use a scope, it must be a real filesystem path containing every
-  changed file.
+  changed file. CI validates this — the title check will fail if any
+  changed file falls outside the scope directory.
 - Use a broader path scope, or omit the scope, for cross-cutting changes.
+  If a PR touches `coderd/`, `codersdk/`, and `site/`, omit the scope
+  entirely: `feat: add per-user threshold overrides`.
 - Example: `fix(coderd/chatd): ...` for changes only in `coderd/chatd/`.
```
Comment (mafredri) [inline AGENTS.md:224]: 🤖 Agent on behalf of mafredri

**P2** Drive-by change to project-level contribution guidelines, unrelated to the feature. *(Luffy P1, Mafu-san P2)*

> The diff adds three lines to the "Code Style" section explaining scope rules and using `feat: add per-user threshold overrides` as the example. The AGENTS.md guidelines say "don't modify code unrelated to your task" and "unnecessary changes make PRs harder to review."

The clarification about CI validation is useful, but it should be a separate PR so it gets its own review. The example also references this PR's own feature (`feat: add per-user threshold overrides`), which will age poorly as a generic style guide example. Could you split this into a follow-up?

File: coderd/chatd/chatd.go [hunk 13]
```diff
@@ -3715,6 +3723,26 @@ func (p *Server) resolveInstructions(
 	return instruction
 }
 
+// resolveUserCompactionThreshold looks up the user's per-model
+// compaction threshold override. Returns the override value and
+// true if one exists and is valid, or 0 and false otherwise.
+func (p *Server) resolveUserCompactionThreshold(ctx context.Context, userID uuid.UUID, modelConfigID uuid.UUID) (int32, bool) {
+	key := codersdk.ChatCompactionThresholdKeyPrefix + modelConfigID.String()
```
Comment (mafredri) [inline coderd/chatd/chatd.go]: 🤖 Agent on behalf of mafredri

**P3** Two maintenance traps in this function: duplicated key construction and hardcoded validation bounds. *(Robin P2, Killua P2, Ryosuke P2 on key dup; Bisky P2, Hisoka P2, Knuckle P2, Takumi P2 on validation range)*

**Key duplication:** `chats.go` has a `compactionThresholdKey(modelConfigID)` helper that does this same `ChatCompactionThresholdKeyPrefix + modelConfigID.String()` concatenation. Two construction sites for the same key format means a future format change needs two edits. Consider exporting a `CompactionThresholdKey` function in `codersdk` next to the prefix constant and using it in both places.

**Hardcoded bounds (line 3740):** `val < 0 || val > 100` uses literals while the API handler in `chats.go` uses `minChatContextCompressionThreshold` / `maxChatContextCompressionThreshold`. If those constants change, this function silently diverges. Worth importing the same constants — or, since the write path already validates, you could trust the stored value and drop the range check entirely here.

Neither is a bug today, but together they create a fragile resolver function.

File: site/src/pages/AgentsPage/UserCompactionThresholdSettings.tsx [hunk 14]
```diff
@@ -0,0 +1,295 @@
+import { getErrorMessage } from "api/errors";
```
Comment (mafredri) [inline site/src/pages/AgentsPage/UserCompactionThresholdSettings.tsx:1]: 🤖 Agent on behalf of mafredri

**P3** No validation feedback when the user types an invalid value — the Save button silently disables. *(Nami P2, Meruem Obs)*

> `parseThresholdDraft` rejects non-integer strings (e.g. "50.5", "1e2") by returning `null`. When `parsedDraftValue === null`, `isSaveDisabled` becomes `true`. But no error message is rendered for this state. A `type="number"` input accepts `.`, `e`, `+`, `-` characters from the keyboard, so a user can easily produce a value like "50.5" that disables Save with no indication of what's wrong.

Consider rendering a brief inline hint (e.g. "Enter a whole number 0–100") when `draftValue` is non-empty but `parsedDraftValue` is null. The 100% warning shows there's already precedent for inline feedback.

PS. Mafuuu also noted the asymmetry: a warning shows for 100% ("will disable compaction") but not for 0% ("will compact every turn"). A parallel hint for very low values could prevent surprise.

File: site/src/api/queries/chats.ts [hunk 15]
```diff
@@ -438,6 +438,35 @@ export const updateUserChatCustomPrompt = (queryClient: QueryClient) => ({
 	},
 });
 
+const userCompactionThresholdsKey = ["user-compaction-thresholds"] as const;
```
Comment (mafredri) [inline site/src/api/queries/chats.ts]: 🤖 Agent on behalf of mafredri

**Nit** Query key `"user-compaction-thresholds"` breaks the `chat-*` prefix convention used by every other key in this file. *(Ging)*

> Every sibling key uses a `chat-` prefix (`chat-system-prompt`, `chat-desktop-enabled`, `chat-user-custom-prompt`, `chat-model-configs`, …). The inconsistency makes bulk invalidation by prefix harder and makes the key harder to find by convention.

Rename to `"chat-user-compaction-thresholds"` for consistency.

File: site/src/pages/AgentsPage/AgentDetail.tsx [hunk 16]
```diff
@@ -231,6 +232,26 @@ export function useConversationEditingState(deps: {
 	};
 }
 
+/**
+ * Resolves the effective compaction threshold for a model configuration,
+ * preferring the user's override when set.
+ */
```
Comment (mafredri) [inline site/src/pages/AgentsPage/AgentDetail.tsx:238]: 🤖 Agent on behalf of mafredri

**Nit** `resolveCompressionThreshold` mixes naming: every other new symbol in this PR uses "compaction." *(Ging, Gon)*

> The backend type is `UserChatCompactionThreshold`, the API path says `user-compaction-thresholds`, the component is `UserCompactionThresholdSettings`, and the server-side resolver is `resolveUserCompactionThreshold`. This one function says "Compression." A reader searching for "compaction" won't find it.

Rename to `resolveCompactionThreshold`. The `compressionThreshold` variable at line 520 can keep its name since it feeds an existing prop, but the helper itself should use the feature's domain term.

**Review by mafredri — DISMISSED** [review]
🤖 Agent on behalf of mafredri

Clean vertical slice — SDK types, SQL, dbauthz wrappers, handlers, client, frontend all present and consistent. Reusing `user_configs` avoids a migration, which is the right call for an experimental per-user preference. The authorization model (`ActionReadPersonal`/`ActionUpdatePersonal`) is correct, the SQL upsert is idempotent, and the test suite covers CRUD, boundaries, validation, and cross-user isolation well.

The main gap is the missing runtime integration test: if the six lines that wire the override into `CompactionOptions` were deleted, every test still passes. That's the one thing I'd want addressed before merge. The rest is a handful of P2/P3 maintenance and consistency items.

1 P1, 3 P2, 2 P3, 2 nits across 8 inline comments.

> Eighteen reviewers, one verdict: test the thing the feature actually does. — Terveisin koodikone.

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review

**Comment by ibetitsmike** [comment]
@codex review
